searchParams.Set panic with "runtime error: slice bounds out of range" if there are >= 3 params with the same name

There is a panic occurred
```
panic: runtime error: slice bounds out of range [3:2]

goroutine 1 [running]:
github.com/nlnwa/whatwg-url/url.(*searchParams).Set(0xc00006c2a0, {0x4c44c1, 0x5}, {0x4c41a2, 0x1})
       .../test-whatwg/whatwg-url/url/searchparams.go:108 +0x425
main.main()
        .../test-whatwg/test/main.go:12 +0x5c
exit status 2
```
if trying to use `Set()` on `searchParams` which contains three or more params with the same name. POC:
```go
package main

import (
	"github.com/nlnwa/whatwg-url/url"
)

func main() {
	u, err := url.Parse("http://test.host/?param=1&param=2&param=3")
	if err != nil {
		panic(err)
	}
	u.SearchParams().Set("param", "4")
}
```

--- a/url/searchparams.go
+++ b/url/searchparams.go
@@ -102,14 +102,18 @@ func (s *searchParams) Has(name string) bool {
 
 func (s *searchParams) Set(name, value string) {
 	isSet := false
-	for idx, nvp := range s.params {
+	idx := 0
+	for _, nvp := range s.params {
 		if nvp.Name == name {
 			if isSet {
 				s.params = append(s.params[:idx], s.params[idx+1:]...)
 			} else {
 				nvp.Value = value
 				isSet = true
+				idx++
 			}
+		} else {
+			idx++
 		}
 	}
 	if !isSet {

--- a/url/searchparams.go
+++ b/url/searchparams.go
@@ -102,22 +102,22 @@ func (s *searchParams) Has(name string) bool {
 
 func (s *searchParams) Set(name, value string) {
 	isSet := false
-	idx := 0
-	for _, nvp := range s.params {
+	params := s.params[:0]
+	for i, nvp := range s.params {
 		if nvp.Name == name {
 			if isSet {
-				s.params = append(s.params[:idx], s.params[idx+1:]...)
-			} else {
-				nvp.Value = value
-				isSet = true
-				idx++
+				s.params[i] = nil
+				continue
 			}
-		} else {
-			idx++
+			nvp.Value = value
+			isSet = true
 		}
+		params = append(params, nvp)
 	}
 	if !isSet {
-		s.Append(name, value)
+		s.params = append(params, &NameValuePair{Name: name, Value: value})
+	} else {
+		s.params = params
 	}
 	s.update()
 }

--- a/url/searchparams_test.go
+++ b/url/searchparams_test.go
@@ -151,6 +151,7 @@ func TestUrlSearchParams_Set(t *testing.T) {
 		{"4", "http://example.com?foo=bar&foo=bar2", "foo", "xyz", "foo=xyz"},
 		{"5", "http://example.com?xyz=aaa&foo=bar2&xyz=aaa&foo=bar", "foo", "xyz", "xyz=aaa&foo=xyz&xyz=aaa"},
 		{"6", "http://example.com?xyz=aaa&foo=bar2&xyz=aaa&foo=bar", "foo2", "xyz", "xyz=aaa&foo=bar2&xyz=aaa&foo=bar&foo2=xyz"},
+		{"7", "http://example.com?foo=bar&foo=fuzz&foo=barfuzz", "foo", "xyz", "foo=xyz"},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {

